### PR TITLE
[NTOS:MM] Fix paged pool initialization on x64

### PR DESCRIPTION
**Pull request by tkreuzer**
Fix paged pool initialization on x64

### CODE REVIEW

File: ntoskrnl/mm/ARM3/mminit.c [hunk 1]
```diff
@@ -1819,12 +1819,12 @@ MiBuildPagedPool(VOID)
...
     // Now check how many PTEs will be required for these many pages.
     //
-    Size = (Size + (1024 - 1)) / 1024;
+    Size = (Size + (PTE_PER_PAGE - 1)) / PTE_PER_PAGE;
 
     //
     // Recompute the page-aligned size of the paged pool, in bytes and pages.
     //
-    MmSizeOfPagedPoolInBytes = Size * PAGE_SIZE * 1024;
+    MmSizeOfPagedPoolInBytes = Size * PAGE_SIZE * PTE_PER_PAGE;
```
Comment (ThFabba) [inline ntoskrnl/mm/ARM3/mminit.c:1827]: This seems strange. Why does "size of paged pool in bytes" depend on how many PTEs fit in a page?

**Review by ThFabba — APPROVED** [review]

File: ntoskrnl/mm/ARM3/mminit.c (hunk 1 above)
Comment (tkreuzer) [reply ntoskrnl/mm/ARM3/mminit.c:1827]: The comments are misleading. Obviously this was written by Alex. And obviously he had no idea, what he was doing 😄 It's not calculating "how many PTEs will be required for these many pages", but how many PDEs are required. And somehow the logic is that paged pool is allocated at PDE granularity (there is an ASSERT below). Don't ask me why, ask Alex, maybe he knows why, but probably not, and probably newer Windows versions don't do this crazy stuff. Long story short: "It's what Windows does". I'll fix the comments.

- [ ] Fix comments.

Comment (tkreuzer) [reply ntoskrnl/mm/ARM3/mminit.c]: I fixed the comments as well as introduce additional variables other than "Size" to make stuff a little bit more readable,

**Review by ThFabba — APPROVED** [review]
Thanks! Much more obvious.

File: ntoskrnl/mm/ARM3/mminit.c [hunk 2]
```diff
@@ -1814,17 +1814,17 @@ MiBuildPagedPool(VOID)
...
     //
-    Size = (Size + (1024 - 1)) / 1024;
+    NumberOfPdes = (NumberOfPages + (PTE_PER_PAGE - 1)) / PTE_PER_PAGE;
 
     //
-    // Recompute the page-aligned size of the paged pool, in bytes and pages.
+    // Recompute the PDE-aligned size of the paged pool, in bytes and pages.
     //
-    MmSizeOfPagedPoolInBytes = Size * PAGE_SIZE * 1024;
+    MmSizeOfPagedPoolInBytes = NumberOfPdes * PTE_PER_PAGE * PAGE_SIZE;
```
Comment (learn-more) [inline ntoskrnl/mm/ARM3/mminit.c:1827]: Using both Pde and PTE here (note: I have no clue if that is a bug or not, it just seems weird)

Comment (tkreuzer) [reply ntoskrnl/mm/ARM3/mminit.c:1827]: Each PDE refers to one page of PTEs. Multiplying number of PDEs with PTEs per page equals total number of PTEs, i.e. total number of pages.

**Pull request merged**